src/coreclr/jit/gentree.cpp

```diff
@@ -25989,6 +25989,18 @@ GenTree* Compiler::gtNewSimdShuffleNode(
     uint64_t value  = 0;
     simd_t   vecCns = {};
 
+    if (simdSize == 16)
+    {
+        // Vector128.Shuffle(a, Vector128.Create(2, 3, 0, 1)) -> ExtractVector128(v.AsUInt64(), v.AsUInt64(), 1)
+        if ((op2->GetIntegralVectorConstElement(0, TYP_ULONG) == 0x300000002) &&
+            (op2->GetIntegralVectorConstElement(1, TYP_ULONG) == 0x100000000))
+        {
+            GenTree* op1Clone = fgMakeMultiUse(&op1);
+            return gtNewSimdHWIntrinsicNode(type, op1, op1Clone, gtNewIconNode(1), NI_AdvSimd_ExtractVector128,
+                                            CORINFO_TYPE_ULONG, simdSize);
```

Comment:
We could also recognize and optimize `Create(GetUpper(), GetLower())` to become the same thing.

Reply:
I likewise wonder if it'd be better to do this optimization in `gtFoldExpr` or `morph` instead, so we can catch cases where the shuffle parameter becomes constant later?

Reply:
> I likewise wonder if it'd be better to do this optimization in `gtFoldExpr` or `morph` instead, so we can catch cases where the shuffle parameter becomes constant later?

My vote for moving all Shuffle optimizations/transformations to lower. E.g. this transformation might produce a local making the tree a bit more complicated, less constant-prop friendly, etc. I'll think about it in a separate PR!